crates/iota-indexer/src/indexer_reader.rs

```diff
@@ -712,10 +707,102 @@ impl IndexerReader {
                 )
                 // we filter the `tx_global_order` table because it is indexed by digest,
                 // optimistic_transactions table is not
-                .filter(tx_global_order::tx_digest.eq_any(missing_digests))
+                .filter(tx_global_order::tx_digest.eq_any(digests))
                 .select(OptimisticTransaction::as_select())
                 .load::<OptimisticTransaction>(conn)
-        })?;
+        })
+    }
+
+    fn get_optimistic_transactions_with_cp_info(
+        &self,
+        digests: &[Vec<u8>],
+    ) -> Result<Vec<(OptimisticTransaction, Option<i64>)>, IndexerError> {
+        run_query!(&self.pool, |conn| {
+            optimistic_transactions::table
+                .inner_join(
+                    tx_global_order::table.on(optimistic_transactions::global_sequence_number
+                        .eq(tx_global_order::global_sequence_number)
+                        .and(
+                            optimistic_transactions::optimistic_sequence_number
+                                .eq(tx_global_order::optimistic_sequence_number),
+                        )),
+                )
+                // we filter the `tx_global_order` table because it is indexed by digest,
+                // optimistic_transactions table is not
+                .filter(tx_global_order::tx_digest.eq_any(digests))
+                .select((
+                    OptimisticTransaction::as_select(),
+                    tx_global_order::chk_tx_sequence_number,
+                ))
+                .load::<(OptimisticTransaction, Option<i64>)>(conn)
+        })
+    }
+
+    async fn get_single_transaction(
+        &self,
+        digest: TransactionDigest,
+    ) -> IndexerResult<Option<StoredTransaction>> {
+        let digest_bytes = digest.inner().to_vec();
+        let mut optimistic_tx_future = {
+            let digest_bytes = digest_bytes.clone();
+            Box::pin(self.spawn_blocking(move |this| {
+                this.get_optimistic_transactions_with_cp_info(&[digest_bytes])
+                    .map(|mut txs| txs.pop())
+            }))
```

Comment:
Nit: consider pinning in the stack with [tokio::pin!](https://docs.rs/tokio/latest/tokio/macro.pin.html)